packages/modules/promotion/src/utils/compute-actions/buy-get.ts

```diff
@@ -97,57 +97,19 @@ export function getComputedActionsForBuyGet(
     return computedActions
   }
 
-  const eligibleItemsByPromotion: EligibleItem[] = []
-  let accumulatedQuantity = MathBN.convert(0)
-
-  /*
-    Eligibility of a BuyGet promotion can span across line items. Once an item has been chosen
-    as eligible, we can't use this item or its partial remaining quantity when we apply the promotion on
-    the target item.
-
-    We build the map here to use when we apply promotions on the target items.
-  */
-
-  for (const eligibleBuyItem of eligibleBuyItems) {
-    if (MathBN.gte(accumulatedQuantity, minimumBuyQuantity)) {
-      break
-    }
-
-    const reservableQuantity = MathBN.min(
-      eligibleBuyItem.quantity,
-      MathBN.sub(minimumBuyQuantity, accumulatedQuantity)
-    )
-
-    if (MathBN.lte(reservableQuantity, 0)) {
-      continue
-    }
-
-    eligibleItemsByPromotion.push({
-      item_id: eligibleBuyItem.id,
-      quantity: MathBN.min(
-        eligibleBuyItem.quantity,
-        reservableQuantity
-      ).toNumber(),
-    })
-
-    accumulatedQuantity = MathBN.add(accumulatedQuantity, reservableQuantity)
-  }
-
-  // Store the eligible buy items for this promotion code in the map
-  eligibleBuyItemMap.set(promotion.code!, eligibleItemsByPromotion)
-
-  // If we couldn't accumulate enough items to meet the minimum buy quantity, return early
-  if (MathBN.lt(accumulatedQuantity, minimumBuyQuantity)) {
-    return computedActions
-  }
-
-  // Get the number of target items that should receive the discount
-  const targetQuantity = MathBN.convert(
+  // Get the number of target items that should receive the discount per application
+  const targetQuantityPerApplication = MathBN.convert(
     promotion.application_method?.apply_to_quantity ?? 0
   )
 
+  // Get the maximum total quantity that can receive the discount across all applications
```

Comment:
though: a lot of changes happened here, and it's not super easy to review ... would maybe be helpful to split this into sub functions, so it is more clear what happens at each stage

Reply:
It changes the application method to now run in a loop, so we now need to keep track of a few more things on each loop. I'm refactoring it now to split them into functions.